### ISSUE TITLE
Non-Null Assertion Operator Is Required When It Doesn't Need to Be

### ISSUE THREAD

**Issue by jvanderen1** [microsoft/TypeScript#50649]
# Bug Report

When I run through an exhaustive check on every `undefined` case to check for non `undefined` values, TS is still required a non-null assertion operator.

### 🔎 Search Terms

- Non-Null Assertion Operator Not Needed

### 🕗 Version & Regression Information

I am currently using TS Playground 4.7.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUA3AjAfgFyJggC2ARgKYBOANIhgEz6EkWUCUiA3gLABQiiGMESYsiALyTE4ACblgMMORmIAZKrr1EAQimz5i5Rx78BiSuSghKSLHwEBfPvcHDROvWDkKlK9ZolPb0MZYxcBCysbRABaO1MnXhchEWxA8WkvA181DQZ0zODfMNMIy2skAAYXRJcAejrEABUAeQARFoIAZQALOBAAGxUwOChCcmVCBBiiAYHEAEMAZyWqWAREOAAHKgX1sGS3NIA+TRKzc3Lo+MdnUwbm9s7EXv6hqbGlSZGwGcH55arSj7TY7Sh7eAHUwpTBaAA8dCw5zMkQqsRuiFqplR0WqvAcQA)

### 💻 Code

```ts
function foo(v1?: number, v2?: number) {
  if (v1 === undefined && v2 !== undefined) {
    return 1
  }

  if (v1 !== undefined && v2 === undefined) {
    return -1
  }

  if (v1 === undefined && v2 === undefined) {
    return 0
  }

  // TODO: Should not need non-null assertion operation
  if (v1 > v2) {
    return 1
  }

  // TODO: Should not need non-null assertion operation
  if (v2 < v1) {
    return -1
  }

  return 0
}
```

### 🙁 Actual behavior

The lines with TODO are complaining about the values of `v1` and `v2` possibly being `undefined`.

### 🙂 Expected behavior

We know the values in this case cannot be `undefined`. That is because there are only 4 possible cases and we have proven in the steps proceeding the TODO's that `undefined` is not possible:

| v1      | v2 | Outcome |
| ----------- | ----------- | --- |
| `undefined`      | `undefined`       |  ❌
| `number`   | `undefined`        | ❌
| `undefined`      | `number`       | ❌
| `number`      | `number`       | ✅

**Comment by jcalz** [comment]
Duplicate of or strongly related to #30869
